common/transport/amqp/src/amqp.ts

```diff
@@ -135,6 +88,280 @@ export class Amqp {
+            if (!this._receivers[endpoint]) {
+              this._fsm.handle('attachReceiverLink', endpoint, null, (err) => {
+                if (err) {
+                  done(err);
+                } else {
+                  done(undefined, this._receivers[endpoint]);
+                }
+              });
+            } else {
+              /*Codes_SRS_NODE_COMMON_AMQP_16_009: [If a receiver for this endpoint has already been created, the getReceiver method should call the done() method with the existing instance as an argument.] */
+              this._safeCallback(done, null, this._receivers[endpoint]);
+            }
+          },
+          attachReceiverLink: (endpoint: string, linkOptions: any, done: GenericAmqpBaseCallback<ReceiverLink>): void => {
+            debug('creating receiver link for: ' + endpoint);
+            this._receivers[endpoint] = new ReceiverLink(endpoint, linkOptions, this._amqp);
+            const attachErrorHandler = (err) => {
+              debug('receiver link error while attaching: ' + endpoint + ': ' + err.toString());
+              delete(this._receivers[endpoint]);
+              done(err);
+            };
+
+            this._receivers[endpoint].on('error', attachErrorHandler);
+            this._receivers[endpoint].attach((err) => {
+              if (err) {
+                debug('failed to attach receiver link: ' + endpoint + ': ' + err.toString());
+                done(err);
+              } else {
+                this._receivers[endpoint].removeListener('error', attachErrorHandler);
+                this._receivers[endpoint].on('error', (err) => {
+                  debug('error on sender link: ' + endpoint + ': ' + err.toString());
+                  delete(this._receivers[endpoint]);
+                });
+                debug('sender link attached: ' + endpoint);
+                done(null, this._receivers[endpoint]);
+              }
+            });
+          },
+          attachSenderLink: (endpoint: string, linkOptions: any, done: GenericAmqpBaseCallback<any>): void => {
+            debug('creating sender link for: ' + endpoint);
+            let senderFsm = new SenderLink(endpoint, linkOptions, this._amqp);
+            this._senders[endpoint] = senderFsm;
+            const attachErrorHandler = (err) => {
+              debug('sender link error event while attaching: ' + endpoint + ': ' + err.toString());
+              delete(this._senders[endpoint]);
+              done(err);
+            };
+
+            this._senders[endpoint].on('error', attachErrorHandler);
+            debug('attaching sender link for: ' + endpoint);
+            this._senders[endpoint].attach((err) => {
+              if (err) {
+                debug('failed to attach sender link: ' + endpoint + ': ' + err.toString());
+                done(err);
```

Comment:
>done [](start = 16, length = 4)

should you delete(this._senders[endpoint]) here?

Reply:
Also, are we guaranteed that the link is not attached?  Do we need to call detach to make absolutely sure that the link is deached?

---
In reply to: [139526430](https://github.com/Azure/azure-iot-sdk-node/pull/83#discussion_r139526430) [](ancestors = 139526430)

Reply:
we should delete. if the link is attached at that point it's a bug in the link state machine.. not convinced we should be that defensive. since error handling and detaching is kinda complicated already i'd rather not do that, might come back to bite us if we ever change the state machine logic of the link?

---
In reply to: [139526967](https://github.com/Azure/azure-iot-sdk-node/pull/83#discussion_r139526967) [](ancestors = 139526967,139526430)